src/main/scala/org/constellation/consensus/EdgeProcessor.scala

```diff
@@ -508,36 +508,50 @@ object Snapshot {
   val snapshotZeroHash: String = Snapshot("", Seq()).hash
 
   def acceptSnapshot(snapshot: Snapshot)(implicit dao: DAO): Unit = {
-    // dao.dbActor.putSnapshot(snapshot.hash, snapshot)
     val cbData = snapshot.checkpointBlocks.map { dao.checkpointService.get }
 
     if (cbData.exists { _.isEmpty }) {
       dao.metrics.incrementMetric("snapshotCBAcceptQueryFailed")
     }
 
-    for (cbOpt <- cbData;
-         cbCache <- cbOpt;
-         cb <- cbCache.checkpointBlock;
-         message <- cb.messages) {
-      dao.messageService.memPool.updateSync(
-        message.signedMessageData.hash,
-        _.copy(snapshotHash = Some(snapshot.hash)),
-        ChannelMessageMetadata(message, Some(cb.baseHash), Some(snapshot.hash))
-      )
-      dao.metrics.incrementMetric("messageSnapshotHashUpdated")
-    }
+    val cbs = cbData.flatten.map(_.checkpointBlock)
+    cbs.foreach { cb =>
+      cb.messagesMerkleRoot.map { messagesMerkleRoot =>
+        val updates = dao.messageService
+          .findHashesByMerkleRoot(messagesMerkleRoot)
+          .map(
+            _.get.map { msgHash =>
+              dao.metrics.incrementMetric("messageSnapshotHashUpdated")
+              dao.messageService.memPool
+                .update(
+                  msgHash,
+                  _.copy(snapshotHash = Some(snapshot.hash)),
+                  ChannelMessageMetadata(
+                    DataResolver
+                      .resolveMessages(msgHash,
+                                       dao.readyPeers.map(p => PeerApiClient(p._1, p._2.client)))
+                      .unsafeRunSync()
```

Comment:
I think you can try to make it simpler, also to remove all `unsafeRunSync()` which are inside (here and below).
It can be written in a more declarative way which will be easier to maintain and will be less error-prone.

Reply:
This one is passed as optional function when key doesn't exists.